File: src/ios/WifiSDKPlugin.m
#import "WifiSDKPlugin.h"
@import GuglielmoConnectSDK;

@implementation WifiSDKPlugin

- (void)initialize:(CDVInvokedUrlCommand *)command {
    NSDictionary *options = [command.arguments objectAtIndex:0];
    NSString *username = options[@"username"];
    NSString *password = options[@"password"];

    if (!username || [username length] == 0 || !password || [password length] == 0) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"username and password are required"];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    [[GuglielmoConnectSDK instance] initializeWithUsername:username password:password];

    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"SDK initialized"];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)isNetworkAdded:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        [[GuglielmoConnectSDK instance] isSuggestedNetworkAlreadyAddedWithCompletion:^(BOOL alreadyAdded, id response) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:alreadyAdded];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }];
    }];
}

- (void)addNetwork:(CDVInvokedUrlCommand *)command {
    NSDictionary *options = [command.arguments objectAtIndex:0];
    NSString *userId = options[@"userId"];
    NSDictionary *additionalInfo = options[@"additionalInfo"];

    if (!userId || [userId length] == 0) {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"userId is required"];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        return;
    }

    [self.commandDelegate runInBackground:^{
        [[GuglielmoConnectSDK instance] askUserToAddSuggestedNetworkWithUserIdentifier:userId additionalInfo:additionalInfo completion:^(BOOL success, id response) {
            if (success) {
                CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:YES];
                [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
            } else {
                CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Failed to add network"];
                [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
            }
        }];
    }];
}

@end

File: src/ios/WifiSDKPlugin.swift
import Foundation
import class GuglielmoConnectSDK.GuglielmoConnectSDK
import struct GuglielmoConnectSDK.ConnectResponse

@objc(WifiSDKPlugin)
class WifiSDKPlugin: CDVPlugin {

    private let sdk = GuglielmoConnectSDK.instance

    @objc(initialize:)
    func initialize(command: CDVInvokedUrlCommand) {
        guard let options = command.arguments?.first as? [String: Any],
              let username = options["username"] as? String, !username.isEmpty,
              let password = options["password"] as? String, !password.isEmpty else {
            let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "username and password are required")
            self.commandDelegate.send(result, callbackId: command.callbackId)
            return
        }

        sdk.initialize(username: username, password: password)

        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "SDK initialized")
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }

    @objc(isNetworkAdded:)
    func isNetworkAdded(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            self.sdk.isSuggestedNetworkAlreadyAdded { alreadyAdded, response in
                let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: alreadyAdded)
                self.commandDelegate.send(result, callbackId: command.callbackId)
            }
        })
    }

    @objc(addNetwork:)
    func addNetwork(command: CDVInvokedUrlCommand) {
        guard let options = command.arguments?.first as? [String: Any],
              let userId = options["userId"] as? String, !userId.isEmpty else {
            let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "userId is required")
            self.commandDelegate.send(result, callbackId: command.callbackId)
            return
        }

        let additionalInfo = options["additionalInfo"] as? [String: String]

        self.commandDelegate.run(inBackground: {
            self.sdk.askUserToAddSuggestedNetwork(
                userIdentifier: userId,
                additionalInfo: additionalInfo
            ) { success, response in
                if success {
                    let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: true)
                    self.commandDelegate.send(result, callbackId: command.callbackId)
                } else {
                    let message = response.message ?? "Failed to add network"
                    let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: message)
                    self.commandDelegate.send(result, callbackId: command.callbackId)
                }
            }
        })
    }
}